### PR TITLE
Use uda value

### PR DESCRIPTION
**Pull request by joakim-hove**
Use the UDA values in the relevant keywords. With this PR the UDA/UDQ work in opm-common is "complete"[1], there is some minor refactoring + MPI (how hard can that be .... :smile:  ) left in opm-simulators. 

#806 is part of this PR.

Have tried hard to do things properly, but some technical debt has undoubtedly been aquired in the process. What stands out is probably the interaction between `DeckItem`, `UDAValue` and `Dimension` - that is currently implemented with some hacky state-changes.

[1]: The quotes on "complete" here is the understatement of the year, the list of not-yet-working UDA/UDQ features is significantly longer than the list of working features, but the feature that now should work is to define your own UDQ values in the SCHEDULE section, and then subsequently use those values in the `WCONPROD` keyword:
```
SCHEDULE

UDQ
     DEFINE 'MY_RATE'  LOG10 'WWCT:OP1' /
/

WCONPROD
   'OP2' .... 'MY_RATE' ... /
/
```
that should be the functionality required for upcoming field studies. 

In the list of not-yet-working features the incomplete support in the keywords different from `WCONPROD` is what stands out the most. For instance for `GCONPROD` the parsing will work with a UDAValue in the deck, but while constructing the `Schedule` object the `double` value of the corresponding `UDAValue` will be accessed unconditionally
```C++
const DeckItem& orat_item = record.getitem("ORAT"); 
double gcon_orat = orat_item.get<double>();
```
and that will blow up with an exception if the user has entered a UDA value. But also for these not-yet-fully supported keywords most of the heavy lifting has been done.

Update: Cooperation with @tskille has relieved some deficits in my understanding of how the UDA values are used; new f...ups will follow!

Working sha:    89465deda

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
OK - this now "runs" the UDQ testcase in opm-tests, but ..... nothing converges and the simulator goes home without completing a single timestep.

**Comment by joakim-hove** [comment]
OK - this now actually seems to work; at least in runs through and unmodified version of the [OPM_WCONDPROD.DATA](https://github.com/OPM/opm-tests/blob/master/udq_actionx/UDQ_WCONPROD.DATA) testcase.

Good with a review - but of course 90% of the UDQ related changes have already been merged.

File: src/opm/output/eclipse/AggregateWellData.cpp [hunk 1]
```diff
@@ -239,7 +237,7 @@ namespace {
             return WMCtrlVal::WMCtlUnk;
         }
 
-        int compOrder(const Opm::Well2& well)
+      int compOrder(const Opm::Well2& well, const Opm::SummaryState& st)
```
Comment (bska) [inline src/opm/output/eclipse/AggregateWellData.cpp]: Why does the static connection ordering need a dynamic `SummaryState`?  As far as I can tell, the `st` parameter is unused in `compOrder()`.

Also, please restore the original indentation level of the function declaration.

File: src/opm/output/eclipse/AggregateWellData.cpp [hunk 2]
```diff
@@ -133,18 +133,17 @@ namespace {
             return ind;
         }
 
-        int wellType(const Opm::Well2& well)
+      int wellType(const Opm::Well2& well, const Opm::SummaryState& st)
```
Comment (bska) [inline src/opm/output/eclipse/AggregateWellData.cpp]: Please restore original indentation level of function declaration.

File: src/opm/output/eclipse/AggregateWellData.cpp [hunk 3]
```diff
@@ -154,22 +153,21 @@ namespace {
             }
         }
 
-        int wellVFPTab(const Opm::Well2& well)
+      int wellVFPTab(const Opm::Well2& well, const Opm::SummaryState& st)
```
Comment (bska) [inline src/opm/output/eclipse/AggregateWellData.cpp]: Please restore original indentation level of function declaration.

File: src/opm/output/eclipse/AggregateWellData.cpp [hunk 4]
```diff
@@ -154,22 +153,21 @@ namespace {
...
             if (well.isInjector()) {
-                return well.injectionControls(summaryState).vfp_table_number;
+                return well.injectionControls(st).vfp_table_number;
             }
-            return well.productionControls(summaryState).vfp_table_number;
+            return well.productionControls(st).vfp_table_number;
         }
 
-        int ctrlMode(const Opm::Well2& well)
+      int ctrlMode(const Opm::Well2& well, const Opm::SummaryState& st)
```
Comment (bska) [inline src/opm/output/eclipse/AggregateWellData.cpp]: Please restore original indentation level of function declaration.

File: src/opm/output/eclipse/AggregateWellData.cpp (hunk 1 above)
Comment (joakim-hove) [reply src/opm/output/eclipse/AggregateWellData.cpp]: > Why does the static connection ordering need a dynamic SummaryState? As far as I can tell, the st parameter is unused in compOrder().

That is right - removed now.

File: src/opm/output/eclipse/AggregateWellData.cpp [hunk 5]
```diff
@@ -304,7 +303,7 @@ namespace {
             // the target control mode requested in the simulation deck.
             // This item is supposed to be the well's actual, active target
             // control mode in the simulator.
-            iWell[Ix::ActWCtrl] = ctrlMode(well);
+            iWell[Ix::ActWCtrl] = ctrlMode(well, st);
```
Comment (bska) [inline src/opm/output/eclipse/AggregateWellData.cpp]: I believe this `st` argument is a left-over from an earlier edition of the patch.

Comment (joakim-hove) [reply src/opm/output/eclipse/AggregateWellData.cpp]: Ehh - no; the callsite you are pointing at now is actually in use.

Comment (bska) [reply src/opm/output/eclipse/AggregateWellData.cpp]: > Ehh - no; the callsite you are pointing at now is actually in use.

Ooops.  Yes, you're right.  I misread `ctrlMode` as `compOrder`.  Sorry.

**Comment by bska** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Ok - I'll merge this now, thank you for the patience with the very last leg of the effort.

Unfortunately there is still work to do, the discrepancy between flow results and Eclipse results are a bit too large, misunderstanding/bug when it comes to *when* UDQ should be evaluated is my current main suspect.

**Pull request merged**